lib/vetext/service.rb

```diff
@@ -0,0 +1,140 @@
+    #
+    # @endpoint_sid String    the registration id as returned from `register`
+    #
+    # @return Hash            response object
+    #
+    def get_preferences(endpoint_sid)
+      response = perform(
+        :get,
+        "#{PREFERENCES_PATH}/#{endpoint_sid}",
+        nil
+      )
+      response.body
+    rescue Common::Client::Errors::ClientError => e
+      remap_error(e)
+    end
+
+    # Set a single preference for a given user/device.
+    #
+    # @endpoint_sid String          the registration id as returned from `register`
+    # @preference_id String         the preference type identifier
+    # @receive_preference boolean   true: user wishes to receive this type of push notification
+    #
+    # @return Hash                  response object
+    #
+    def set_preference(endpoint_sid, preference_id, receive_preference)
+      response = perform(
+        :put,
+        PREFERENCES_PATH, {
+          endpointSid: endpoint_sid,
+          preferenceId: preference_id,
+          value: receive_preference == true
+        }
+      )
+      raise_if_response_error(response.body)
+      response.body
+    rescue Common::Client::Errors::ClientError => e
+      remap_error(e)
+    end
+
+    # Send a push notification to a single device
+    #
+    # @endpoint_sid String    the registration id as returned from `register`
+    # @template_id String     id of the push notification content template
+    # @personalisation Hash   data map provided by sender to fill in specified template
+    #
+    # @return Hash            response object
+    #
+    def send_notification(endpoint_id, template_id, personalization = {})
+      response = perform(
+        :post,
+        SEND_PATH, {
+          endpointSid: endpoint_id,
+          templateSid: template_id,
+          personalization: personalization
+        }
+      )
+      raise_if_response_error(response.body)
+      response.body
+    rescue Common::Client::Errors::ClientError => e
+      remap_error(e)
+    end
+
+    private
+
+    # Raise an error if the service returned an error in the
+    # body of a 200 response
+    #
+    def raise_if_response_error(body)
+      raise VEText::ResponseError, body if body[:success] == false
+    end
+
+    def remap_error(e)
+      case e.status
+      when 400..499
+        raise Common::Exceptions::BackendServiceException.new('VETEXT_PUSH_400',
+                                                              { detail: e.body[:error] }, e.status,
+                                                              e.body)
+      when 500..599
+        raise Common::Exceptions::BackendServiceException.new('VETEXT_PUSH_502', {}, e.status)
+      else
+        raise e
+      end
+    end
+
+    def get_app_sid(app_name)
+      settings = Settings.vetext_push
+      if settings.key?("#{app_name}_sid".to_sym)
+        settings["#{app_name}_sid".to_sym]
+      else
+        raise Common::Exceptions::BackendServiceException, 'VETEXT_PUSH_404'
```

Comment:
I wouldn't consider this a VETEXT error, since we're generating it ourselves. I think we can raise a standard error, and return a 400 from the controller.